File: backend/app/schemas/imports.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict

from app.models.imports import Currency, SheetType, StockType


class ImportItemParsed(BaseModel):
    ean: str
    product_name: str
    price: float


class PreviewMetadata(BaseModel):
    suggested_header_index: int | None
    suggested_mapping: dict[str, int | None] | None
    header_columns: dict[str, str] | None


class PreviewResponse(BaseModel):
    filename: str
    metadata: PreviewMetadata
    raw_grid: list[list[Any]]


class ImportBatchResponse(BaseModel):
    id: int
    filename: str
    supplier_name: str
    sheet_type: str
    stock_type: str
    currency: str
    description: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ImportItemResponse(BaseModel):
    id: int
    ean: str
    product_name: str
    price: float

    comparison_price: float | None = None
    comparison_supplier: str | None = None
    comparison_batch_id: int | None = None
    comparison_currency: str | None = None
    margin_percentage: float | None = None

    model_config = ConfigDict(from_attributes=True)


class ImportBatchCreateResponse(BaseModel):
    message: str
    batch_id: int
    rows_inserted: int


class ImportBatchUpdate(BaseModel):
    supplier_name: str | None = None
    description: str | None = None
    sheet_type: SheetType | None = None
    stock_type: StockType | None = None
    currency: Currency | None = None


class MatchSourceItem(BaseModel):
    item_id: int
    ean: str
    product_name: str
    price: float
    currency: str
    sheet_type: str


class MatchItem(BaseModel):
    item_id: int
    product_name: str
    price: float
    batch_id: int
    batch_filename: str
    supplier_name: str
    currency: str
    margin_percentage: float | None = None


class ItemMatchesResponse(BaseModel):
    source: MatchSourceItem
    matches: list[MatchItem]

File: frontend/src/types/imports.ts
export type SheetType = "buy" | "sell";
export type StockType = "stock" | "preorder";
export type Currency = "EUR" | "USD" | "GBP" | "CZK";

export interface PaginatedResponse<T> {
  data: T[];
  total: number;
  skip: number;
  limit: number;
}

export interface ImportBatch {
  id: number;
  filename: string;
  supplier_name: string;
  sheet_type: SheetType;
  stock_type: StockType;
  currency: Currency;
  description: string | null;
  created_at: string;
}

export interface ImportItem {
  id: number;
  ean: string;
  product_name: string;
  price: number;
  comparison_price: number | null;
  comparison_supplier: string | null;
  comparison_batch_id: number | null;
  comparison_currency: string | null;
  margin_percentage: number | null;
}

export interface PreviewMetadata {
  suggested_header_index: number | null;
  suggested_mapping: Record<string, number | null> | null;
  header_columns: Record<string, string> | null;
}

export interface PreviewResponse {
  filename: string;
  metadata: PreviewMetadata;
  raw_grid: unknown[][];
}

export interface BatchCreateResponse {
  message: string;
  batch_id: number;
  rows_inserted: number;
}

export interface BatchUpdatePayload {
  supplier_name?: string;
  description?: string | null;
  sheet_type?: SheetType;
  stock_type?: StockType;
  currency?: Currency;
}

export interface MatchSourceItem {
  item_id: number;
  ean: string;
  product_name: string;
  price: number;
  currency: string;
  sheet_type: SheetType;
}

export interface MatchItem {
  item_id: number;
  product_name: string;
  price: number;
  batch_id: number;
  batch_filename: string;
  supplier_name: string;
  currency: string;
  margin_percentage: number | null;
}

export interface ItemMatchesResponse {
  source: MatchSourceItem;
  matches: MatchItem[];
}

export interface BatchCreatePayload {
  file: File;
  supplier_name: string;
  sheet_type: SheetType;
  stock_type: StockType;
  currency: Currency;
  description: string;
  header_index: number;
  ean_col: number;
  product_col: number;
  price_col: number;
}
